onnxscript/optimizer/_inliner.py

```diff
@@ -0,0 +1,272 @@
+            attributes = {**attributes, **default_attr_values}
+
+        if len(node.inputs) > len(function.inputs):
+            raise ValueError(f"Input mismatch: {len(node.inputs)} > {len(function.inputs)}")
+        value_map = {}
+        for i, input in enumerate(node.inputs):
+            value_map[function.inputs[i]] = input
+        for i in range(len(node.inputs), len(function.inputs)):
+            value_map[function.inputs[i]] = None
+
+        # Identify call-stack for node, used to generate unique names.
+        call_stack = self.node_context.get(node, [])
+        call_stack.append(call_site_id)
+
+        cloner = _CopyReplace(self, attributes, value_map, node.metadata_props, call_stack)
+
+        # iterate over the nodes in the function, creating a copy of each node
+        # and replacing inputs with the corresponding values in the value map.
+        # Update the value map with the new values.
+
+        nodes = [cloner.clone_node(node) for node in function]
+        output_values = [value_map[output] for output in function.outputs]
+        return nodes, output_values  # type: ignore
+
+    def inline_calls_in(self, graph: ir.Graph) -> None:
+        for input in graph.inputs:
+            if input.name is not None:
+                self.used_value_names.add(input.name)
+        for initializer in graph.initializers:
+            self.used_value_names.add(initializer)
+
+        # Pre-processing:
+        # * Count the number of times each function is called in the graph.
+        #   This is used for disambiguating names of values in the inlined functions.
+        # * And identify names of values that are used in the graph.
+        id_count : dict[ir.OperatorIdentifier, int] = defaultdict(int)
+        for node in graph:
+            if node.name:
+                self.used_node_names.add(node.name)
+            id = node.op_identifier()
+            if id in self._functions:
+                id_count[id] += 1
+            for output in node.outputs:
+                if output.name is not None:
+                    self.used_value_names.add(output.name)
+        next_id : dict[ir.OperatorIdentifier, int] = defaultdict(int)
+        for node in graph:
+            id = node.op_identifier()
+            if id in self._functions:
+                # If there are multiple calls to same function, we use a prefix to disambiguate
+                # the different call-sites:
+                if id_count[id] > 1:
+                    call_site_prefix = f"_{next_id[id]}"
+                    next_id[id] += 1
+                else:
+                    call_site_prefix = ""
+                call_site = node.name or (self._function_id_abbreviations[id] + call_site_prefix)
+                nodes, values = self._instantiate_call(node, call_site)
+                convenience.replace_nodes_and_values(
+                    graph, node, [node], nodes, node.outputs, values
+                )
+            else:
```

Comment:
Does this assume a function call will not take graph attributes?

Reply:
Good question. I don't think this is relevant to the code here in the else branch, which is not inside a function. But the question is relevant to what happens in the code in CopyReplace, when we do substitution of attribute values for attribute parameters. I think that code does assume that a function call will not take graph attributes for now ... I think it will cause an assertion failure (in some cases). Generalizing to handle that would be interesting ... doable, I think, but the complexity is mostly in ensuring correctness.